### PR TITLE
zipkinv2: add encoders, decoders, marshalers

### PR DESCRIPTION
**Pull request by jrcamp**
Also update zipkin receiver to use them.

### CODE REVIEW

File: receiver/kafkareceiver/zipkin_unmarshaler.go [hunk 1]
```diff
@@ -32,16 +29,13 @@ type zipkinProtoSpanUnmarshaler struct {
...
 
 var (
-	_            TracesUnmarshaler = (*zipkinProtoSpanUnmarshaler)(nil)
-	toTranslator                   = zipkinv2.ToTranslator{ParseStringTags: false}
+	_                         TracesUnmarshaler = (*zipkinProtoSpanUnmarshaler)(nil)
+	zipkinProtobufUnmarshaler                   = zipkinv2.NewProtobufTracesUnmarshaler(false, false)
+	zipkinJSONUnmarshaler                       = zipkinv2.NewJSONTracesUnmarshaler(false)
 )
 
 func (z zipkinProtoSpanUnmarshaler) Unmarshal(bytes []byte) (pdata.Traces, error) {
```
Comment (bogdandrutu) [inline receiver/kafkareceiver/zipkin_unmarshaler.go:37]: Use https://github.com/open-telemetry/opentelemetry-collector/blob/main/receiver/kafkareceiver/pdata_unmarshaler.go.

File: receiver/kafkareceiver/zipkin_unmarshaler_test.go [hunk 2]
```diff
@@ -113,13 +113,13 @@ func TestUnmarshalZipkinThrift_error(t *testing.T) {
 func TestUnmarshalZipkinJSON_error(t *testing.T) {
 	p := zipkinJSONSpanUnmarshaler{}
 	got, err := p.Unmarshal([]byte("+$%"))
-	assert.Equal(t, pdata.NewTraces(), got)
+	assert.Equal(t, pdata.Traces{}, got)
```
Comment (bogdandrutu) [inline receiver/kafkareceiver/zipkin_unmarshaler_test.go:116]: No need to compare that is invalid, if error we should not use the pdata result. Probably remove this line.

File: receiver/kafkareceiver/zipkin_unmarshaler_test.go [hunk 3]
```diff
@@ -113,13 +113,13 @@ func TestUnmarshalZipkinThrift_error(t *testing.T) {
...
 	assert.Error(t, err)
 }
 
 func TestUnmarshalZipkinProto_error(t *testing.T) {
 	p := zipkinProtoSpanUnmarshaler{}
 	got, err := p.Unmarshal([]byte("+$%"))
-	assert.Equal(t, pdata.NewTraces(), got)
+	assert.Equal(t, pdata.Traces{}, got)
```
Comment (bogdandrutu) [inline receiver/kafkareceiver/zipkin_unmarshaler_test.go:123]: ditto

File: receiver/zipkinreceiver/trace_receiver.go [hunk 4]
```diff
@@ -132,30 +133,20 @@ func (zr *ZipkinReceiver) v2ToTraceSpans(blob []byte, hdr http.Header) (reqs pda
...
-
-	default: // By default, we'll assume using JSON
-		zipkinSpans, err = zr.deserializeFromJSON(blob)
-	}
-
-	if err != nil {
-		return pdata.Traces{}, err
+	if hdr.Get("Content-Type") == "application/x-protobuf" {
+		// TODO: (@odeke-em) record the unique types of Content-Type uploads
+		if debugWasSet {
```
Comment (bogdandrutu) [inline receiver/zipkinreceiver/trace_receiver.go:142]: Can you please investigate the reason why this was added and maybe consider to remove it if no longer useful.

**Review by bogdandrutu — APPROVED** [review]

**Comment by tigrannajaryan** [comment]
Please resolve conflicts.

**Comment by tigrannajaryan** [comment]
Please resolve the conflicts.

File: translator/trace/zipkinv2/protobuf.go [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+	"go.opentelemetry.io/collector/internal/model"
+)
+
+var _ model.TracesDecoder = (*protobufDecoder)(nil)
+
+type protobufDecoder struct {
+	// DebugWasSet toggles the Debug field of each Span. It is usually set to true if
+	// the "X-B3-Flags" header is set to 1 on the request.
+	DebugWasSet bool
```
Comment (rakyll) [inline translator/trace/zipkinv2/protobuf.go:29]: Hah, very cryptic name but I agree that we should use what zipkin is using.

File: receiver/zipkinreceiver/trace_receiver.go (hunk 4 above)
Comment (rakyll) [reply receiver/zipkinreceiver/trace_receiver.go:142]: Zipkin uses debug=true to indicate that sampling decisions is overriden but I'm not sure why it is recorded as a part of the span. I assume, it allows developers to query traces initiated for debugging purposes. It's a part of their span model.

**Review by rakyll — APPROVED** [review]

File: receiver/zipkinreceiver/trace_receiver.go (hunk 4 above)
Comment (jrcamp) [reply receiver/zipkinreceiver/trace_receiver.go:140]: @bogdandrutu does this resolve it?

**Comment by jrcamp** [comment]
@bogdandrutu updated, lint failure looks unrelated

File: receiver/zipkinreceiver/trace_receiver.go (hunk 4 above)
Comment (bogdandrutu) [reply receiver/zipkinreceiver/trace_receiver.go:140]: That is for in process propagation as far as I know, not for collection. This is where it is unclear to me

File: translator/trace/zipkinv2/protobuf.go (hunk 5 above)
Comment (bogdandrutu) [reply translator/trace/zipkinv2/protobuf.go:29]: Also why public in a private structure?

**Pull request merged**